test/negativebinomial.jl

```diff
@@ -1,20 +1,35 @@
 using Distributions
+using ChainRulesTestUtils
+using FiniteDifferences
 using Test, ForwardDiff
 
-# Currently, most of the tests for NegativeBinomail are in the "ref" folder.
+# Currently, most of the tests for NegativeBinomial are in the "ref" folder.
 # Eventually, we might want to consolidate the tests here
 
 mydiffp(r, p, k) = r/p - k/(1 - p)
 
-@testset "NegativeBinomial r=$r, p=$p, k=$k" for
-    p in exp10.(-10:0) .- eps(), # avoid p==1 since it's not differentiable
-        r in exp10.(range(-10, stop=2, length=25)),
-            k in (0, 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024)
+@testset "NegativeBinomial r=$r" for r in exp10.(range(-10, stop=2, length=25))
+    # avoid p==1 since it's not differentiable
+    @testset "p = $p" for p in exp10.(-10:0) .- eps()
+        @testset "k = $k" for k in (0, 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024)
+            @test ForwardDiff.derivative(_p -> logpdf(NegativeBinomial(r, _p), k), p) ≈ mydiffp(r, p, k) rtol=1e-12 atol=1e-12
 
-    @test ForwardDiff.derivative(_p -> logpdf(NegativeBinomial(r, _p), k), p) ≈ mydiffp(r, p, k) rtol=1e-12 atol=1e-12
+            dist = NegativeBinomial(r, p)
+            if p == 1. - eps()
+                # Did not get to work:
+                # test_rrule(logpdf, dist, k; fdm = backward_fdm(5, 1, max_range = r/2), rtol = 1e-5, atol = 1e-1)
+            else
+                test_rrule(logpdf, dist, k; fdm = central_fdm(5, 1, max_range = min(r, p)/2), rtol = 1e-10, atol = 1e-1)
```

Comment:
The tolerances seem suspicious. `rtol` is smaller than the default (1e-6 or 1e-9, I don't remember?) but `atol` is very large. It's difficult to say if that would catch any issues without knowing the order of magnitude of the values (derivatives, primals, etc.).

Can we add tests with the default tolerances of 1e-9? The ForwardDiff tests above even use 1e-12. Maybe finite differencing is just too inaccurate for some of the extreme values here (e.g., `exp10(-10)`), so maybe it's just not possible to check derivatives for all these parameters.

In principle, a single, possibly random set of parameters r and p and sample k for each of the branches would be sufficient: one where k is in the support, one where k is not in the support, and one where p = 1 and k  = 0.

Reply:
I agree. I initially tuned these to make everything pass. So those are the minimium tolerances that can work for all parameters (to within one order ox magnitude). To decrease, we have to drop certain parameters from the test, or change the FDM, but I haven't found one yet that allows for significantly lower tolerances.

I wonder, since the ForwardDiff tests check against the analytical derivative, and tge rrule is analytically derived, can we just check the latter for a few random parameters? It does not seem that the finite difference tests accomplish much here.

Reply:
I would prefer to use finite differencing in the tests - you don't want to use it in practice due to the numerical issues and inaccuracies but it's the simplest way to compute derivatives, and hence the one least likely to be systematically incorrect. ForwardDiff derivatives can be incorrect in some special cases but it can be tricky to notice and debug, and analytical derivatives can be wrong and were used to define the rules.

So checking some random, not too extreme parameters with finite differencing would be the best, I think.